draft-ietf-mls-protocol.md

```diff
@@ -4092,3 +4091,118 @@ def common_ancestor_direct(x, y, _):
        k += 1
     return (xn << k) + (1 << (k-1)) - 1
 ~~~~~
+
+
+# Extending and Truncating Link-Based Trees
+
+An implementation may choose to store ratchet trees in a "link-based"
+representation, where each node stores references to its parents and/or
+children.   (As opposed to the array-based representation suggested above, where
+these relationships are computed from relationships between nodes' indices in
+the array.)  Such an implementation needs to update these links to maintain the
+left-balanced structure of the tree as the tree is extended to add new members,
+or truncated when memebers are removed.
+
+This appendix provides example algorithms for adding a single leaf and removing
+a single leaf from a link-based tree.  These algorithms can be repeated as
+necessary to implement the tree modifications required to apply Add and Remove
+proposals.
+
+To add a leaf node N to the right side of the tree:
```

Comment:
Do we need both, pseudo code and python code? I'd be fine with the python code only and some of this as comments if you feel like it.